### PR TITLE
[WIP] Connect v2

### PR DESCRIPTION
**Pull request by dzucconi**
connect-view

In progress... actual code is in https://github.com/aredotna/ervell/commit/6db262f7c8320ebfe4e3b8fb607d6b85c05db1ee

I'm sort of undecided about this but I think what I'm going to do for integration and less confusing refactoring is sub-classing the existing views I'm integrating with: https://github.com/aredotna/ervell/commit/f80e8aace5482b7d38d8470e7261c8b3f2f07c0c#diff-3434079a7960bae6217d6ef6ba4f72eeR5

(I've decided to leave collection/model stuff alone for now)

### CODE REVIEW

File: components/connect_v2/client/index.coffee [hunk 1]
```diff
@@ -1,22 +1,30 @@
...
 
-  $channels
-    .on 'click', (e) ->
-      e.preventDefault()
+  else
+    Promise.resolve()
 
-      $(this).attr 'data-selected'
+  replenish.then ->
+    channels.fetch(data: per: 3 - channels.length) if channels.length < 0
```
Comment (dzucconi) [inline components/connect_v2/client/index.coffee]: Just plumbing while I work on this

File: apps/block/client/view.coffee [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{ extend, defer, invoke } = require 'underscore'
+LegacyBlockView = require './legacy_view.coffee'
```
Comment (broskoski) [inline apps/block/client/view.coffee]: hahahha, just so you don't have to look at it.

File: components/connect_v2/client/index.coffee (hunk 1 above)
Comment (broskoski) [inline components/connect_v2/client/index.coffee]: Actually not sure what is going on here, the `recent_connections` in `mediator` should be coming from localStorage. Why replenish them?

File: components/connect_v2/client/view.coffee [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ConnectResultsView = require '../components/results/view.coffee'
+
+module.exports = class ConnectView extends Backbone.View
+  className: 'Connect'
+
+  attributes:
+    'data-state': 'inactive'
+
+  initialize: ->
+    @subViews = [
```
Comment (broskoski) [inline components/connect_v2/client/view.coffee]: 🎨

File: components/connect_v2/client/index.coffee (hunk 1 above)
Comment (dzucconi) [inline components/connect_v2/client/index.coffee]: Ah yeah. I was trying to transliterate what existed in the view already. It's really confusing though. Will revisit.

Comment (broskoski) [inline components/connect_v2/client/index.coffee]: Happy to pair, currently on adding a connection it pops it into the a collection of recent connections (stored in local storage).

Comment (dzucconi) [inline components/connect_v2/client/index.coffee]: Nah I get that. Basically I just need to not execute this line if there's a length. The code is still very weird.

**Comment by broskoski** [comment]
This looks beau, is this round going to have the "new private channel" functionality in https://github.com/aredotna/ervell/issues/519 ?

File: components/connect_v2/client/index.coffee (hunk 1 above)
Comment (dzucconi) [inline components/connect_v2/client/index.coffee]: https://github.com/aredotna/ervell/blob/e383c80c3585e1cb2d61abc4325b530f6d5ee0bc/components/connect/client/connect_view.coffee#L117-L123

**Comment by dzucconi** [comment]
Yeah sure, what the heck

File: components/connect_v2/client/index.coffee (hunk 1 above)
Comment (broskoski) [inline components/connect_v2/client/index.coffee]: Right yeah, so `fetch` in this context ( https://github.com/jeromegn/Backbone.localStorage ) means fetch from localStorage.

Comment (dzucconi) [inline components/connect_v2/client/index.coffee]: (which is a deeply weird API choice)

Comment (broskoski) [inline components/connect_v2/client/index.coffee]: I suppose it's due to the fact that getting items from localStorage is asynchronous.

Comment (broskoski) [inline components/connect_v2/client/index.coffee]: The number of results should also probably be configurable (mockups show 5 results + new channel option).

Comment (dzucconi) [inline components/connect_v2/client/index.coffee]: I was actually about to "fix" this. This is literally the identical code in the existing view. Note the `if channels.length < 0`

**Comment by dzucconi** [comment]
This wraps up the implementation.

connecting

I'm going to hit the integration now and then get it ready for merge. BRB moving to a cafe.

File: components/connect_v2/client/index.coffee [hunk 4]
```diff
@@ -1,4 +1,5 @@
 ConnectView = require './view.coffee'
+config = require '../config.coffee'
```
Comment (broskoski) [inline components/connect_v2/client/index.coffee]: Nice

File: components/connect_v2/client/view.coffee [hunk 5]
```diff
@@ -11,27 +11,30 @@ module.exports = class ConnectView extends Backbone.View
...
-    @listenTo @searchView, 'reset', @deactivate
-    @listenTo @searchView, 'query', @activate
+    @listenTo @state, 'change:active', @toggle
 
-  activate: ->
-    @$el.attr 'data-state', 'active'
-
-  deactivate: ->
-    @$el.attr 'data-state', 'inactive'
+  toggle: (_state, active) ->
```
Comment (broskoski) [inline components/connect_v2/client/view.coffee]: ✅

File: components/connect_v2/components/item/view.coffee [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          connectable_id: @connectable.id
+          connectable_type: @connectable.get('base_class')
+
+  disconnect: (channel) ->
+    mediator.shared.recent_connections.unshove channel
+
+    mediator.trigger 'connection:removed', channel
+    mediator.trigger "connection:removed:#{@connectable.id}", channel
+
+    @queue.enqueue () =>
```
Comment (dzucconi) [inline components/connect_v2/components/item/view.coffee]: I'm vaguely surprised I never thought to do this before for detaching optimistic UI from the syncing.

UI responds instantly and anything async just gets enqueued. The it runs everything first-in-first-out order. Each channel in the list gets it's own queue so different channel connections run in parallel. Toggling the same channel is serial.

File: components/connect_v2/components/item/view.coffee [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        url: "#{API_URL}/channels/#{channel.id}/connections"
+        data:
+          connectable_id: @connectable.id
+          connectable_type: @connectable.get('base_class')
+
+  disconnect: (channel) ->
+    mediator.shared.recent_connections.unshove channel
+
+    mediator.trigger 'connection:removed', channel
+    mediator.trigger "connection:removed:#{@connectable.id}", channel
```
Comment (dzucconi) [inline components/connect_v2/components/item/view.coffee]: Mediator stuff gets triggered immediately, however

File: lib/async_serial_queue.coffee [hunk 8]
```diff
@@ -0,0 +1,20 @@
+module.exports = class AsyncSerialQueue
```
Comment (broskoski) [inline lib/async_serial_queue.coffee]: o la la

File: components/connect_v2/components/item/view.coffee (hunk 6 above)
Comment (broskoski) [inline components/connect_v2/components/item/view.coffee]: Yeah, good idea. I suppose the next step would be to abort a particular queue in the event of an API error?

**Comment by dzucconi** [comment]
@broskoski OK I think this is done and integrated now. I'll merge tomorrow morning so we can take a look on staging

**Comment by broskoski** [comment]
🍾 

Very exciting 💵 💵 💵 

Shall dig through in a bit.

File: apps/block/client/view.coffee [hunk 9]
```diff
@@ -11,19 +11,18 @@ module.exports = class BlockView extends LegacyBlockView
...
     e.preventDefault()
 
-    $el = @$('.block-connect-container')
-    $el.addClass 'is-active'
+    $target = $(e.currentTarget)
 
-    view = new ConnectView
-      el: $el
-      block: @model
+    view = new InlineConnectIntegrationView model: @model
```
Comment (broskoski) [inline apps/block/client/view.coffee]: 👍

File: apps/block/templates/info/_connections.jade [hunk 10]
```diff
@@ -14,8 +14,5 @@
     //- Rendered client-side
 
   if user
-    .block-sidebar__section--connect__button.pointer.js-connect
-      | Connect &#8594;
-
-  .block-connect-container.connect-container
-    //- Rendered client-side
+    .Button.Button--size-sm.js-connect
```
Comment (broskoski) [inline apps/block/templates/info/_connections.jade]: 👍

File: components/block_collection/client/block_view.coffee [hunk 11]
```diff
@@ -34,34 +34,34 @@ module.exports = class BlockView extends Backbone.View
...
+    @listenTo @model, 'remote_update', @update
+    @listenTo @model, 'show', @show
+    @listenTo @model, 'hide', @hide
 
-    mediator.on "model:#{@model.id}:updated", @update, @
-    mediator.on "connection:#{@model.id}:complete", @removeActiveClass, @
-    mediator.on "body:click", @removeActiveClass, @
+    @listenTo mediator, "model:#{@model.id}:updated", @update
+    @listenTo mediator, "connection:#{@model.id}:complete", @removeActiveClass
+    @listenTo mediator, "body:click", @removeActiveClass
```
Comment (broskoski) [inline components/block_collection/client/block_view.coffee]: thank ye

File: components/block_collection/stylesheets/block_collection.styl [hunk 12]
```diff
@@ -396,4 +396,5 @@ body.is-mobile
     display static !important
 
 @require '../../lightbox/stylesheets/lightbox'
-@require '../../connect/stylesheets/connect'
+@require '../../connect/stylesheets/connect' // This includes some really random styles unrelated to actual connect, so we still need it for now... :\
```
Comment (broskoski) [inline components/block_collection/stylesheets/block_collection.styl]: sorryyyy

**Comment by broskoski** [comment]
Very nice, can't wait to see 👌

**Pull request merged**